src/VarArray.mo

```diff
@@ -107,104 +223,740 @@ module {
+                  element
+                };
+                case null {
+                  Prim.trap "Malformed array in mapResults"
+                }
+              }
+            }
+          )
+        )
+      };
+      case (?error) {
+        error
+      }
+    }
   };
 
-  public func mapEntries<T, Y>(array : [var T], f : (T, Nat) -> Y) : [var Y] {
-    todo()
+  /// Creates a new array by applying `f` to each element in `array` and its index.
+  /// Retains original ordering of elements.
+  ///
+  /// ```motoko include=import
+  ///
+  /// let array = [10, 10, 10, 10];
+  /// Array.mapEntries<Nat, Nat>(array, func (x, i) = i * x)
+  /// ```
+  ///
+  /// Runtime: O(size)
+  ///
+  /// Space: O(size)
+  ///
+  /// *Runtime and space assumes that `f` runs in O(1) time and space.
+  public func mapEntries<T, R>(array : [var T], f : (T, Nat) -> R) : [var R] {
+    tabulate<R>(array.size(), func i = f(array[i], i))
   };
 
-  public func flatMap<T, R>(array : [var T], k : T -> [var R]) : [var R] {
-    todo()
+  /// Creates a new array by applying `k` to each element in `array`,
+  /// and concatenating the resulting arrays in order.
+  ///
+  /// ```motoko include=import
+  /// import Nat "mo:base/Nat";
+  ///
+  /// let array = [var 1, 2, 3, 4];
+  /// VarArray.flatMap<Nat, Int>(array, func x = [x, -x])
+  ///
+  /// ```
+  /// Runtime: O(size)
+  ///
+  /// Space: O(size)
+  /// *Runtime and space assumes that `k` runs in O(1) time and space.
+  public func flatMap<T, R>(array : [var T], k : T -> Iter.Iter<R>) : [var R] {
+    var flatSize = 0;
+    let arrays = Prim.Array_tabulate<[var R]>(
+      array.size(),
+      func i {
+        let subArray = fromIter<R>(k(array[i])); // TODO: optimize
+        flatSize += subArray.size();
+        subArray
+      }
+    );
+
+    // could replace with a call to flatten,
+    // but it would require an extra pass (to compute `flatSize`)
+    var outer = 0;
+    var inner = 0;
+    tabulate<R>(
+      flatSize,
+      func _ {
+        while (inner == arrays[outer].size()) {
+          inner := 0;
+          outer += 1
+        };
+        let element = arrays[outer][inner];
+        inner += 1;
+        element
+      }
+    )
   };
 
+  /// Collapses the elements in `array` into a single value by starting with `base`
+  /// and progessively combining elements into `base` with `combine`. Iteration runs
+  /// left to right.
+  ///
+  /// ```motoko include=import
+  /// import {add} "mo:base/Nat";
+  ///
+  /// let array = [var 4, 2, 0, 1];
+  /// let sum =
+  ///   VarArray.foldLeft<Nat, Nat>(
+  ///     array,
+  ///     0, // start the sum at 0
+  ///     func(sumSoFar, x) = sumSoFar + x // this entire function can be replaced with `add`!
+  ///   );
+  /// ```
+  ///
+  /// Runtime: O(size)
+  ///
+  /// Space: O(1)
+  ///
+  /// *Runtime and space assumes that `combine` runs in O(1) time and space.
   public func foldLeft<T, A>(array : [var T], base : A, combine : (A, T) -> A) : A {
```

Comment:
It's worth checking foldLeft and foldRight have consistent argument across all our collections.

Reply:
Looks like they're currently all consistent. That's a good point knowing that some languages and libraries will swap the arguments for `foldRight()`, which I'd prefer to avoid.